### PR TITLE
server: reduce "growslice" in "dumpRowValuesBinary"

### LINKED ISSUE #4915: Improve the performance of function dumpRowValuesBinary
I see a lot of growslice in this function.

### PR DESCRIPTION
**Pull request by zz-jason**
fix #4915

### CODE REVIEW

File: server/util.go [hunk 1]
```diff
@@ -260,6 +259,8 @@ func dumpRowValuesBinary(alloc arena.Allocator, columns []*ColumnInfo, row []typ
 			nulls[bytePos] |= 1 << bitPos
 		}
 	}
+	data = make([]byte, 0, 1+nullsLen+8*len(row))
+	data = append(data, mysql.OKHeader)
 	data = append(data, nulls...)
```
Comment (tiancaiamao) [inline server/util.go]: How about alloc `data` and `nulls` in slice to avoid this operation?

Comment (zz-jason) [reply server/util.go]: good idea

File: server/util.go [hunk 2]
```diff
@@ -251,17 +251,16 @@ func dumpRowValuesBinary(alloc arena.Allocator, columns []*ColumnInfo, row []typ
 		return
 	}
 	nullsLen := ((len(columns) + 7 + 2) / 8)
-	nulls := make([]byte, nullsLen)
+	data = make([]byte, 1+nullsLen, 1+nullsLen+8*len(row))
```
Comment (ngaut) [inline server/util.go]: Can simply use data = append(mysql.OKHeader, nulls...)

**Comment by shenli** [comment]
@coocood PTAL

**Comment by coocood** [comment]
I think we should change all the `dump` functions to append input []bytes and return it, like the functions in `util/codec`.

**Comment by zimulala** [comment]
LGTM

File: server/conn.go [hunk 3]
```diff
@@ -810,6 +810,8 @@ func (cc *clientConn) writeResultset(rs ResultSet, binary bool, more bool) error
 		return errors.Trace(err)
 	}
 
+	numBytes4Null := ((len(columns) + 7 + 2) / 8)
+	buffer := make([]byte, 1+numBytes4Null, 1+numBytes4Null+8*(len(columns)))
```
Comment (coocood) [inline server/conn.go]: How about `rowBuffer`?

File: server/util.go [hunk 4]
```diff
@@ -245,75 +245,72 @@ func uniformValue(value interface{}) interface{} {
...
 		switch val.Kind() {
 		case types.KindInt64:
 			v := val.GetInt64()
 			switch columns[i].Type {
 			case mysql.TypeTiny:
-				data = append(data, byte(v))
+				buffer = append(buffer, byte(v))
 			case mysql.TypeShort, mysql.TypeYear:
-				data = append(data, dumpUint16(uint16(v))...)
+				buffer = append(buffer, dumpUint16(uint16(v))...)
```
Comment (coocood) [inline server/util.go]: How about:
`buffer = dumpUint16(buffer, uint16(v))`

File: server/util.go [hunk 5]
```diff
@@ -245,75 +245,72 @@ func uniformValue(value interface{}) interface{} {
...
 			floatBits := math.Float32bits(val.GetFloat32())
-			data = append(data, dumpUint32(floatBits)...)
+			buffer = append(buffer, dumpUint32(floatBits)...)
 		case types.KindFloat64:
 			floatBits := math.Float64bits(val.GetFloat64())
-			data = append(data, dumpUint64(floatBits)...)
+			buffer = append(buffer, dumpUint64(floatBits)...)
 		case types.KindString, types.KindBytes:
-			data = append(data, dumpLengthEncodedString(val.GetBytes(), alloc)...)
+			buffer = append(buffer, dumpLengthEncodedString(val.GetBytes(), alloc)...)
```
Comment (coocood) [inline server/util.go]: `buffer = dumpLengthEncodedString(buffer, val.GetBytes())`
Then we don't need to use `alloc`.

File: server/util.go [hunk 6]
```diff
@@ -245,75 +245,72 @@ func uniformValue(value interface{}) interface{} {
 	}
 }
 
-func dumpRowValuesBinary(alloc arena.Allocator, columns []*ColumnInfo, row []types.Datum) (data []byte, err error) {
+func dumpRowValuesBinary(alloc arena.Allocator, columns []*ColumnInfo, row []types.Datum, buffer []byte) ([]byte, error) {
```
Comment (coocood) [inline server/util.go]: I think put `buffer` to the first argument is better.
`alloc` can be removed if all the `dump` methods use `buffer` to append the value.

**Comment by coocood** [comment]
LGTM

**Pull request merged**